Feature: Allow PR status checks to be managed external to safe settings

jitran opened the pull request:
### Background
There have been a number of requests from users who have implemented branch protection rules and rulesets across their organisations, but would like their PR status checks to be defined outside of safe settings.

When these rules are defined at the repository or suborg level, it's not so bad, as the safe settings PR status checks only affect a subset of repositories. Users have a bit more control over their safe settings configuration.

When the rules are defined at the org level, it's a bit more problematic, as the safe settings PR status checks affect all repositories in the org.

### Changes
- Added the ability for branch protection rules and rulesets' status checks to be managed outside of safe settings.
- When users specify `{{EXTERNALLY_DEFINED}}` in the status checks for their branch protection rules or rulesets, safe settings will not revert status checks in GitHub.
- Created an Overrides module to handle externally defined settings.
- Updated usage documentation in the README.
- Added unit tests for the new functionality.

### Tests
Ensured all unit tests pass.

#### Manual testing with rulesets (repo, suborg, and org levels)
The following was tested in sequential order:
1. Started with no rulesets in GitHub for a repository
1. Created a new safe setting for a ruleset that had `{{EXTERNALLY_DEFINED}}`defined  in the status checks
   - Ruleset was deployed with an empty list of status checks
1. Added custom status checks to the new ruleset via GitHub UI
   - The ruleset's custom status checks in GitHub were preserved
1. Updated the ruleset's status checks via the GitHub UI
   - The ruleset's  custom status checks in GitHub were preserved
1. Added a new status check to the existing ruleset in safe settings and left `{{EXTERNALLY_DEFINED}}` in the list
   - The ruleset's  custom status checks in GitHub were preserved
1. Removed all status checks from the ruleset via the GitHub UI
   - The ruleset's empty status checks in GitHub was preserved
1. Removed `{{EXTERNALLY_DEFINED}}` from the existing ruleset's status checks in safe settings
   - The ruleset's status checks in GitHub was reverted to the status checks from safe settings
1. Removed all status checks from the existing ruleset in safe settings
   - The ruleset's status checks in GitHub was reverted to the status checks from safe settings
1. Removed the ruleset from GitHub
   - Safe settings restored the ruleset
1. Added custom status checks to the ruleset via GitHub UI
   - The ruleset's status checks was reverted to the status checks from safe settings
1. Added a completely new ruleset (branch|tag|push) via GitHub UI
   - Safe settings removed all new rulesets

#### Manual testing with branch protection rules (repo, suborg, and org levels)
The following was tested in sequential order:
1. Deleted all branch protection rules from a repository
1. Created two branch protection rules in safe settings; the first rule had `{{EXTERNALLY_DEFINED}}` in the status checks, and the second rule had two status checks defined
   - Two branch protection rules were deployed into GitHub, where the first had no status checks, and the second had two status checks defined.
1. Updated the status checks for the first rule via GitHub UI
   - The custom status checks for the first rule in GitHub were preserved
1. Updated the status checks for the second rule via GitHub UI
   - The custom status checks for the second rule in GitHub were reverted back to the safe settings

jitran committed:
Allow branch protection settings to be overridden outside of safe settings.

jitran committed:
Allow ruleset settings to be overridden outside of safe settings.

jitran committed:
Created Overrides module for handling externally defined settings.

jitran committed:
On top level ruleset updates, it would trigger a sync all that will trigger top level branch protection rules across all repos. Make a clone of the branch settings before modifying it since it's a shared object.

jitran committed:
Updated readme.

jitran committed:
Added unit tests.

jitran committed:
Added unit tests for organisation rulesets.

jitran committed:
Updated tests to reflect latest ruleset data type changes.

PendaGTP commented on README.md:
```diff
@@ -123,6 +123,42 @@ overridevalidators:
+
+This can be defined at the org, sub-org, and repo level.
+
+To configure this for branch protection rules, specifiy `{{EXTERNALLY_DEFINED}}` under the `contexts` keyword:
```
`specifiy` -> `specify`

jitran committed:
Updated documentation.

igorcosta reviewed (changes requested):
@jitran , Thanks for the detailed work on this feature! I’ve reviewed the changes, and overall, this is a **solid** implementation that brings much-needed flexibility to managing PR status checks. That said, I have a few suggestions to improve error handling, documentation clarity, and test coverage:

- In `branches.js` and `rulesets.js`, there are scenarios where API failures (e.g., 500-series errors) are logged but not surfaced to the user. Consider adding retry logic for transient errors and ensuring failures are gracefully handled with actionable error messages.

- Include a warning about potential desync when rulesets are modified directly via the GitHub UI.
- In `overrides.js`, it would be good to add safeguards in removeOverrides to handle malformed configurations or unexpected data structures more robustly.
- Add an example showing how {{EXTERNALLY_DEFINED}} interacts with hierarchical rulesets (org, repo, and sub-org levels).

Let me know your thoughts on these suggestions! 


Nice job!

jitran committed:
Updated ruleset overrides based on the latest api change.

jitran committed:
Updated documentation and added example scenarios.

jitran commented:
Hi @igorcosta and @PendaGTP, I have rebased my branch with the latest default branch and validated that all the existing tests are still successful.

As for https://github.com/github/safe-settings/pull/741#pullrequestreview-2561095903:
- The documentation has been updated with a warning about potential de-syncs when rulesets are modified directly via the GitHub UI.
- Safe guards have been added to overrides.js to handle configurations and data structures more robustly.
- I have added examples of how `{{EXTERNALLY_DEFINED}}` interacts across Org, Sub-org, and Repo levels.
- I think a separate pull request should be raised to address the transient errors to keep this PR clean though :)

Finally, I had to rework the solution to handle the recent ruleset changes where it no longer allows the `Require status checks to pass` checkbox to be enabled with no status checks defined. It has added further complexity to the solution though.

igorcosta reviewed (approved):
LGTM thanks for revisiting and improving @jitran

powerman commented:
Any chance `{{EXTERNALLY_DEFINED}}` is broken when using safe-settings as an action `npm run full-sync`? In my case it always reset checks to `[]`, not just on first run, but on every run.
